mdbook-tera-backend/src/tera_renderer/renderer.rs

```diff
@@ -0,0 +1,158 @@
+                .ok_or_else(|| tera::Error::from(format!("Could not find key {key} in config")))?;
+            let value = to_value(value)?;
+            Ok(value)
+        }
+    }
+
+    pub(crate) fn render_book(&mut self) -> Result<()> {
+        let dest_dir = &self
+            .ctx
+            .destination
+            .parent()
+            .ok_or_else(|| {
+                RendererError::InvalidPath(format!(
+                    "Destination directory {:?} has no parent",
+                    self.ctx.destination
+                ))
+            })?
+            .to_owned();
+        if !dest_dir.is_dir() {
+            return Err(RendererError::InvalidPath(format!(
+                "{:?} is not a directory",
+                dest_dir
+            )));
+        }
+        self.render_book_directory(&dest_dir)
+    }
+
+    fn render_book_directory(&mut self, path: &Path) -> Result<()> {
+        for entry in path.read_dir()? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_dir() {
+                self.render_book_directory(&path)?;
+            } else {
+                self.process_file(&path)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn process_file(&mut self, path: &Path) -> Result<()> {
+        if path.extension().unwrap_or_default() != "html" {
+            return Ok(());
+        }
+        let mut file_content = String::new();
+        {
+            let mut file = fs::File::open(path)?;
+            file.read_to_string(&mut file_content)?;
+        }
+
+        let output = self.render_components(&file_content, path)?;
+        let mut output_file = fs::File::create(path)?;
+        output_file.write_all(output.as_bytes())?;
+        Ok(())
+    }
+
+    fn render_components(&mut self, file_content: &str, path: &Path) -> Result<String> {
+        let rendering_context = RenderingContext::new(
+            path.to_owned(),
+            self.ctx.config.book.language.clone(),
+            &self.serialized_ctx,
+            &self.ctx,
+        )?;
+        let custom_components_handlers = self
+            .components
+            .iter()
+            .map(|component| {
+                element!(component.component_name(), |el| {
+                    let attributes: BTreeMap<String, String> = el
+                        .attributes()
+                        .iter()
+                        .map(|attribute| (attribute.name(), attribute.value()))
+                        .collect();
+                    let rendered = component.render(&rendering_context, attributes)?;
+                    el.replace(&rendered, ContentType::Html);
+                    Ok(())
+                })
+            })
+            .collect();
+        let output = lol_html::rewrite_str(
+            file_content,
+            RewriteStrSettings {
+                element_content_handlers: custom_components_handlers,
+                ..RewriteStrSettings::default()
+            },
+        )?;
```

Comment:
I bet using a `lol_html` library makes it hard to use this renderer to process other kinds of files (text files)?

If you instead first let the `html` renderer output to a directory, then it will copy over all other files too. Then you can iterate over all `.html`, `.txt`, `.js`, ... files and run them through the template engine.

Reply:
It's a feature we can add in the future without much difficulty, just don't use HTML for non html files based on the extension. That could be useful for JS, but I'm not sure it's something we need